### PR TITLE
[CBR-60] Wallet spec: tracking metadata (transaction history)

### PR DESCRIPTION
**Pull request by edsko**
This describes how the wallet can be extended to track metadata, and how this can be instantiated to provide the transaction history required by the wallet V1 API.

### CODE REVIEW

**Comment by edsko** [comment]
_Copy and pasting here for ease of reference._

@edsko you don't have to support the "Creating" status. It was created as a hack. There's no fundamental reason for it to exist if we don't want it.
It was created as a hack to resolve some problem with something being non-atomic.
"The empty circle means that the transaction is not included in the wallet’s reported history." huhh? what empty circle
sorry I don't yet understand this section
I don't understand the WontApply state, or how we get there
perhaps it's because I have a headache, but I think this needs to be spelled out more clearly
ok, so we can get from applying to wontapply by a transaction being removed from the pending but not added to the chain (since it is invalidated by another tx that does get added to the chain)
so perhaps we might rename this state as "invalidated"?
though once we have TTL we'll also get into this state from the timeout
@edsko I don't think it's a good design to just punt re-submission to the network layer to handle
in the network layer, broadcasting transactions is necessarily somewhat lossy
so I think keeping track of when to re-submit should be done in the wallet
either we keep a little bit of state to track when we last submitted, or we keep track of in which slot we added it to pending, and then just use a function of the time between the current slot and the slot it was added in to decide when to re-submit
ie we have a function that is given the time since the tx was added to pending, in slots, and it decides for each slot number, is this a good one in which to re-submit
resubmitNow :: SlotNo -> Bool
and we can make that return True for 0, so that we always queue new pending txs to be sent when they're first added
so I think we can ask for the network layer to provide us a queue-like interface for queueing txs to broadcast now/soon
and we simply add to that queue each slot when the resubmitNow tells us to
though perhaps we can do slightly better, since that requires a linear scan of the pending set each slot
and plausibly it's better to handle the initial enqueue as a special case, to do it promptly and not on a slot boundary

**Comment by edsko** [comment]
Ok, rephrased and added detail. Not sure if this is better. Pending meeting with @dcoutts .

**Comment by edsko** [comment]
Ok, discussed and will push another commit.

**Comment by edsko** [comment]
Ok, ready for review again and (from my pov) ready to be merged.

**Comment by edsko** [comment]
Rebased on `develop`.

File: wallet-new/docs/spec.tex [hunk 1]
```diff
@@ -2261,17 +2275,50 @@ \subsubsection{Transaction status}
...
+from $\mathit{pending}$ (without being included in the blockchain), perhaps
+because it was invalidated by another transaction (or the transaction submission
+layer gave up on it; see Section~\ref{sec:transaction_submission}). We will
+report a \texttt{WontApply} status for such a transaction.
+\item There is however a second way that we might end up with such an
+outgoing transaction: it may have been created by \emph{another instance} of the
+same wallet. Reporting such a transaction as \texttt{WontApply} is perhaps
+somewhat confusing, as it was never in the \texttt{Applying} state in
+\emph{this} wallet. However, it \emph{was} of course in \texttt{Applying} state
+in the other way, and reporting the transaction as \texttt{WontApply}  also in
```
Comment (dcoutts) [inline wallet-new/docs/spec.tex]: "in the other way" ? do you mean in the other wallet?

**Comment by edsko** [comment]
Fixed typo, rebased on `develop`. Let's merge when green.

**Pull request merged**